common/define-grammar.js

```diff
@@ -504,25 +511,40 @@ module.exports = function defineGrammar(dialect) {
         $._type
       ),
 
-      _primary_type: $ => choice(
+      _conditionable_type: $ => choice(
         $.parenthesized_type,
         $.predefined_type,
         $._type_identifier,
         $.nested_type_identifier,
         $.generic_type,
-        $.type_predicate,
         $.object_type,
         $.array_type,
         $.tuple_type,
-        $.flow_maybe_type,
         $.type_query,
         $.index_type_query,
         $.this,
         $.existential_type,
         $.literal_type,
-        $.lookup_type
+        $.lookup_type,
+        $.conditional_type,
+        $.flow_maybe_type
+      ),
+
+      _primary_type: $ => choice(
+        $._conditionable_type,
+        $.type_predicate,
```

Comment:
I'm not opposed to this new `_conditionable_type` layer, but could you briefly explain why it is needed, as opposed to allowing `type_predicate` to remain part of `primary_type`?

Reply:
`type_predicate`'s usage is more restricted than a general type. As an example: [TypeScript Playground](https://www.typescriptlang.org/play?#code/AQehwEwSwZwQwDYIPYHcCmEBQAXAngA7rAAqwAvMABrCzACawA-NcAFwNZajjTxJpMuQsQDiFVnXpcsAMwCuAOwDGOKMkXA4MGOgBOOAJIxRAHlEA+ABQBzDqICUHG7RjBxAb27A96HPL1NHD15dCwAXyA)

The only official resource I have at the ready is the chapter on assertion functions: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions

Anyways, effectively, they have special treatment. As there were other rules already using this definition of `_primary_type` (I'm not sure it's sound everywhere, but I started from this assumption), I opted for keeping `_primary_type`'s behavior as is currently and introduce this sub-rule as a means to not deal with moving parts.

Another approach, in the opposite direction, would be to extract `type_predicate` out of `_primary_type` and have it be something like `choice($._primary_type, $.type_predicate)` in the places it applies - currently, to my understanding, that'd be in function signature position. This is less straightforward to figure out if it'd be a worthwhile change. It's also less straightforward because it incites the question "is type assertion is an actual type?".

All things considered, the last approach might be better since they have special treatment. I'll try it in another PR.